x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -364,13 +425,16 @@ private void loadApiKeyAndValidateCredentials(ThreadContext ctx, ApiKeyCredentia
                         response.getSourceAsBytesRef(), XContentType.JSON)) {
                         apiKeyDoc = ApiKeyDoc.fromXContent(parser);
                     }
-                    validateApiKeyCredentials(docId, apiKeyDoc, credentials, clock, ActionListener.delegateResponse(listener, (l, e) -> {
-                        if (ExceptionsHelper.unwrapCause(e) instanceof EsRejectedExecutionException) {
-                            listener.onResponse(AuthenticationResult.terminate("server is too busy to respond", e));
-                        } else {
-                            listener.onFailure(e);
+                    if (checkpoint != null) {
+                        final CachedApiKeyDoc cachedApiKeyDoc = apiKeyDoc.toCachedApiKeyDoc();
+                        if (checkpoint.put(docId, cachedApiKeyDoc)) {
+                            roleDescriptorsBytesCache.computeIfAbsent(
+                                cachedApiKeyDoc.roleDescriptorsHash, k -> apiKeyDoc.roleDescriptorsBytes);
+                            roleDescriptorsBytesCache.computeIfAbsent(
+                                cachedApiKeyDoc.limitedByRoleDescriptorsHash, k -> apiKeyDoc.limitedByRoleDescriptorsBytes);
```

Comment:
As a possible refactoring in the future, the logic that puts and gets an API key doc, which uses a secondary cache for the role descriptors is a nice thing to encapsulate.

Reply:
Add a new ApiKeyDocCache inner class to encapsulate the secondary cache. It is the right thing to do for this PR. Thanks!